DT-340: Warn the user if they are leaving a story edit page with unsaved changes

torresga opened the pull request:
### Jira Ticket 
https://ombulabs.atlassian.net/browse/DT-340

### Motivation / Context
Currently if you leave the current page, you can lose information that was not saved. This feature warns you if you are leaving without saving information.
    
### QA / Testing Instructions
1. Go to the Edit Story page.
2. Edit the story.
3. Try to navigate off the page by reloading, pressing the back button in the browser, clicking the logo or the Back button on the page.
4. Ensure that you get a popup asking you if are sure you want to reload/go back.

### Screenshots:

<img width="1071" alt="Screenshot 2024-11-07 at 1 53 28 PM" src="https://github.com/user-attachments/assets/629253ae-fc55-442b-b643-43a164282df5">

___

I will abide by the [code of conduct](https://github.com/fastruby/points/blob/main/CODE_OF_CONDUCT.md).

torresga committed:
Adds event listeners to listen for unsaved edits and alerting the user if there are unsaved edits

torresga committed:
Adds event listeners to listen for unsaved edits and alerting the user if there are unsaved edits

torresga committed:
Adds spec to test for presence of alert when we try to navigate away from an unsaved edited story

torresga committed:
Adds spec to test for presence of alert when we try to navigate away from an unsaved edited story

torresga committed:
Refactored the code to only add the event listeners on the form if the form existed. Also created a function to add and remove the beforeunload event listener

torresga committed:
Refactored the code to only add the event listeners on the form if the form existed. Also created a function to add and remove the beforeunload event listener

JuanVqz commented on app/assets/javascripts/stories.js:
```diff
@@ -25,8 +25,57 @@ document.addEventListener("DOMContentLoaded", () => {
+    element.addEventListener("click", function (event) {
+      if (isDirty) {
+        const confirmLeave = confirm("You have unsaved changes. Are you sure you want to go back?");
+        if (!confirmLeave) {
+          // Prevent navigation if the user chooses not to leave
+          event.preventDefault();
+        } else {
+          // Optionally, reset isDirty if leaving
+          isDirty = false;
+          addBeforeUnloadEventListener(isDirty)
+        }
```
I find it easier to read `confirmLeave` compared to `!confirmLeave`, as it always takes me a moment to grasp the condition.

```suggestion
        if (confirmLeave) {
          // Optionally, reset isDirty if leaving
          isDirty = false;
          addBeforeUnloadEventListener(isDirty)
        } else {
          // Prevent navigation if the user chooses not to leave
          event.preventDefault();
        }
```

JuanVqz commented on app/assets/javascripts/stories.js:
```diff
@@ -25,8 +25,57 @@ document.addEventListener("DOMContentLoaded", () => {
+  }
+}
+
+function warnUserifUnsavedEdits(event) {
```
A bit picky, you might missed the `I` should go in uppercase.
 
```suggestion
function warnUserIfUnsavedEdits(event) {
```

JuanVqz reviewed (commented):
My suggestions are not a blocker.

torresga committed:
DT-340: Edit tests so that it tests all scenarios

torresga committed:
DT-340: Edit tests so that it tests all scenarios

torresga committed:
Update app/assets/javascripts/stories.js

Co-authored-by: Juan Vásquez <juan@ombulabs.com>

torresga committed:
Update app/assets/javascripts/stories.js

Co-authored-by: Juan Vásquez <juan@ombulabs.com>

torresga committed:
Update app/assets/javascripts/stories.js

Co-authored-by: Juan Vásquez <juan@ombulabs.com>

torresga committed:
Update app/assets/javascripts/stories.js

Co-authored-by: Juan Vásquez <juan@ombulabs.com>

JuanVqz committed:
Fix method name

JuanVqz committed:
Fix method name

JuanVqz commented:
Hey @torresga, this looks good; however, maybe we can improve a little bit on the `isDirty` logic; for example, I tried the following, and the state was still dirty when it probably should not have been.

1. Go to the edit page.
2. Type 3 letters in the title.
3. Click back, the logo, or any action that triggers the modal to alert us we left unsaved changes.
4. The alert will pop up as expected; until here, all looks good.
5. Then I removed the 3 letters to be in an **unchanged state**, and clicking any action to get out of the page does not allow me because it says I have unsaved changes when I really don't.

If you need help to follow the described steps, let me know, and we can check it together.

JuanVqz committed:
Fix unsaved-changes detection to compare against initial form state

- Track dirtiness by comparing the current form serialization to a snapshot
  taken on load, so reverting an edit (typing then deleting it) correctly
  clears the warning instead of leaving the form permanently 'dirty'.
- Also listen for 'change' so status-select edits are detected, and guard the
  back/logo click handlers against missing elements on non-edit pages.
- Add a feature spec covering the revert-to-original case.

JuanVqz committed:
Unify unsaved-changes wording and init on turbolinks:load

- Route every in-app link that leaves the edit page through one confirm with a
  single shared message constant, instead of hard-coding only Back and the
  logo. The native beforeunload prompt (reload / tab close) still uses the
  browser's own text, since browsers ignore custom strings there.
- Initialize on turbolinks:load instead of DOMContentLoaded, matching
  project.js, so the guard (and the markdown preview) also initialize on
  Turbolinks visits. Module-scoped state lets the delegated click handler be
  added by reference and de-duplicated across visits.

JuanVqz committed:
Fix flaky clone sub-projects spec ordering

Assert the cloned sub-projects by set of titles (contain_exactly) instead of
positional index. The projects association has no default order, so indexing
into it was order-dependent and failed on the build-rails-next lane when the
database returned the rows in a different order.
